File: MoonEngine-Root/Base/src/main/java/com/greentree/engine/moon/base/assets/scene/Context.java
package com.greentree.engine.moon.base.assets.scene;

import com.greentree.commons.reflection.info.TypeInfo;
import com.greentree.commons.reflection.info.TypeInfoBuilder;
import com.greentree.commons.xml.XMLElement;

public interface Context {

    default <T> T build(Class<T> cls, XMLElement element) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return build(type, element);
    }

    default <T> T build(TypeInfo<T> type, XMLElement element) {
        try (final var c = newInstance(type, element)) {
            if (c == null)
                throw new NullPointerException("type:" + type + " element:" + element);
            return c.value();
        }
    }

    <T> Constructor<T> newInstance(TypeInfo<T> type, XMLElement element);

    default <T> Constructor<T> newInstance(Class<T> cls, XMLElement element) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return newInstance(type, element);
    }

}

File: MoonEngine/Base/src/main/java/com/greentree/engine/moon/base/assets/scene/adapters/Context.kt
package com.greentree.engine.moon.base.assets.scene.adapters

import com.greentree.commons.reflection.ClassUtil
import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.commons.reflection.info.TypeInfoBuilder
import com.greentree.commons.xml.XMLElement

fun <T> findClass(baseClass: Class<T>, className: String): Class<T> {
	@Suppress("UNCHECKED_CAST")
	return ClassUtil.loadClassInAllPackages(baseClass, className) as Class<T>
}

interface Context {

	fun <T : Any> build(cls: Class<T>, element: XMLElement): T {
		val type = TypeInfoBuilder.getTypeInfo(cls)
		return build(type, element)
	}

	fun <T : Any> build(type: TypeInfo<T>, element: XMLElement): T {
		newInstance(type, element).use { c ->
			return c.value()
		}
	}

	fun <T : Any> newInstance(type: TypeInfo<T>, element: XMLElement): Constructor<T>
	fun <T : Any> newInstance(cls: Class<T>, element: XMLElement): Constructor<T> {
		val type = TypeInfoBuilder.getTypeInfo(cls)
		return newInstance(type, element)
	}
}